odl/discr/tensor_ops.py

```diff
@@ -369,7 +369,104 @@ def derivative(self, vf):
         return PointwiseInner(self.domain, inner_vf, weight=self.weights)
 
 
-class PointwiseInner(PointwiseTensorFieldOperator):
+class PointwiseInnerBase(PointwiseTensorFieldOperator):
+    """ Base class for `PointwiseInner` and `PointwiseInnerAdjoint`.
+
+    Implemented to allow code reuse between the classes.
+    """
+
+    def __init__(self, adjoint, vfspace, vecfield, weight=None):
+        """Initialize a new instance.
+
+        All parameters are given according to the specifics of the "usual"
+        operator. The ``adjoint`` parameter is used to control conversions
+        for the inverse transform.
+
+        Parameters
+        ----------
+        adjoint : `bool`
+            True if the operator should be the adjoint, False else.
+        vfspace : `ProductSpace`
+            Space of vector fields on which the operator acts.
+            It has to be a product space of identical spaces, i.e. a
+            power space.
+        vecfield : domain `element-like`
+            Vector field with which to calculate the point-wise inner
+            product of an input vector field
+        weight : `array-like` or `float`, optional
+            Weighting array or constant for the norm. If an array is
+            given, its length must be equal to ``domain.size``, and
+            all entries must be positive. A provided constant must be
+            positive.
+            By default, the weights are is taken from
+            ``domain.weighting``. Note that this excludes unusual
+            weightings with custom inner product, norm or dist.
+        """
+        if not isinstance(vfspace, ProductSpace):
+            raise TypeError('`vfsoace` {!r} is not a ProductSpace '
+                            'instance'.format(vfspace))
+        if adjoint:
+            super().__init__(domain=vfspace[0], range=vfspace, linear=True)
+        else:
+            super().__init__(domain=vfspace, range=vfspace[0], linear=True)
+
+        # Bail out if the space is complex but we cannot take the complex
+        # conjugate.
+        if (vfspace.field == ComplexNumbers() and
+                not hasattr(self.base_space.element_type, 'conj')):
+            raise NotImplementedError(
+                'base space element type {!r} does not implement conj() '
+                'method required for complex inner products'
+                ''.format(self.base_space.element_type))
+
+        self._vecfield = vfspace.element(vecfield)
```

Comment:
Tried keeping the changes to places where we may expect someone to use a private variable by misstake, there are about 500 cases where we could do this change so I'm currently abit lazy in these specific classes.